### DISCUSSION TITLE
What is the possibility of running it on Windows?

### DISCUSSION THREAD

**Discussion by ricardoapaes** [php/frankenphp#420] [General]
What is the possibility of running it on Windows?

I have a use case where I need, in some cases, to have PHP servers on Windows due to the clients' own structure.

I really liked the project and I would like to know if this possibility exists and if this is in the plans or if you could give me an initial push with tips on how to get there, I am interested in helping to make this possible.

**Comment by withinboredom** [comment]
I spent a lot of time working this out in #83 a while ago. PHP builds Windows ZTS builds, but we'd need to compile them from scratch (to add some build flags. I'd have to dig through my notes to know exactly which ones). It is quite a journey, but I'm sure it is possible.

To compile a Windows binary of FrankenPHP, we would also need to compile a custom PHP binary and distribute both together with any extensions (as I'm not sure the prebuilt extensions for Windows will be compatible). If we would need to also distribute extensions, I'm not sure it will be feasable.

I never got it to work, primarily due to not having a deep understanding of the Windows C build chain, caddy, Frankenphp, and Go in Windows. My knowledge of Caddy and Frankenphp has grown massively over the last year, but I still need to gain a basic understanding of how to build PHP in Windows.

If anyone has this knowledge and wants to work together to figure out how to get Frankenphp working in Windows, I'd love to hop on a zoom call (or discord, or whatever) and figure this out.

**Reply by withinboredom** [reply]
I've managed to figure out how to get it sorta-kinda-working in a windows docker container. My shortcomings with making it anywhere too far:

1. I have very little experience with Windows containers.
2. I have no idea if people actually use Windows containers.
3. I have found so many isolation issues with Windows containers that I would be well and truly shocked if people were actually using Windows containers.
4. I fubar'd my Windows installation working on this. twice.

I'm probably not going to continue working on this, however, I'll see if I can clean it up and add some comments so maybe someone else can pick this up.

**Comment by pierrejoye** [comment]
Hello :)

I gave some attempts earlier, mainly for the static build of php on Windows. This is not supported at all by the build scripts (blame me, we did not consider it as relevant back then when I rewrote the whole thing for 5.3+).

So far I can build the libphp as static lib (and the default extensions when using --dsiable-all), but there are other issues with the symbols. I think it is better to build the SAPI statically linked to the libphp and the extension, instead of libphp having the extension statically linked. I will try more :)

Also it may be good to also first get the SAPI built (that should not be a big issue). and work (no idea if there are specific unix-like features we need to emulate, but as it is used by go, that should work but the thread-mutex lib) as DLL first.

- To achieve this, porting C-Thread-Pool to windows is required (https://github.com/dunglas/frankenphp/blob/main/C-Thread-Pool/)
- Create the FrankenPHP as DLL
- Add the cgo flags&co to load and use it

It will help a lot to deal with the static build of PHP on windows (which is quite a tasks as many extensions needs some patching for minit/mshutdown etc to avoid wrong init/shutdown and other shared mem bugs).

**Comment by dunglas** [comment]
This may be huge step forward for Windows support: https://github.com/crazywhalecc/static-php-cli/pull/292

**Reply by ricardoapaes** [reply]
It appears that version 2.0 has been released and the aforementioned pull request is merged. But I didn't find where I can test the version on Windows because it doesn't make executables available in the release (like it did for Mac and Linux).

[EDIT]
I just found documentation on how to compile for Windows, I'll do my tests here.
https://static-php.dev/en/guide/build-on-windows.html

**Comment by aolko** [comment]
Any progress?

**Comment by HassanDomeDenea** [comment]
static-php-cli supports Windows in latest releases.
Any timeline for windows support by frankenphp ?

**Reply by ahmadkakarr** [reply]
Anyone working on it..?
I tried with xcaddy but did not worked?

**Reply by dunglas** [reply]
I'm not aware of someone actively working on this. Contribs welcome!

**Comment by TenHian** [comment]
#880 Maybe I should move it here.

**Comment by ScottHuangZL** [comment]
Do core team have any plan to release FrankenPhp.exe for **Windows**? Thanks.

**Reply by withinboredom** [reply]
There are people working on it, off and on.

In any case, as soon as we can, we will. If this is something that interests you, we need more contributors in this area.

**Reply by Lea-Reift** [reply]
> If this is something that interests you, we need more contributors in this area.

@withinboredom  interests me. What are the exact requeriments to be a contributor?

**Reply by withinboredom** [reply]
@Lea-Reift - PRs are welcome!

**Comment by henderkes** [comment]
Next release will have Windows binaries.

Keep in mind that php itself runs much better on Unix than on Windows, there's nothing we can do about that. I'd still recommend WSL.
